MoreLinq/EndsWith.cs

```diff
@@ -29,21 +29,24 @@ static partial class MoreEnumerable
         /// <typeparam name="T">Type of elements.</typeparam>
         /// <param name="first">The sequence to check.</param>
         /// <param name="second">The sequence to compare to.</param>
+        /// <param name="comparer">Optional equality comparer, defaults to <see cref="EqualityComparer{T}.Default" />.</param>
         /// <returns>
         ///     <c>true</c> if <paramref name="first" /> ends with elements equivalent to <paramref name="second" />.
         /// </returns>
         /// <remarks>
         ///     This is the <see cref="IEnumerable{T}" /> equivalent of <see cref="string.EndsWith(string)" /> and
-        ///     it calls <see cref="object.Equals(object)" /> on pairs of elements at the same index.
+        ///     it calls <see cref="IEqualityComparer{T}.Equals(T,T)" /> on pairs of elements at the same index.
         /// </remarks>
-        public static bool EndsWith<T>(this IEnumerable<T> first, IEnumerable<T> second)
+        public static bool EndsWith<T>(this IEnumerable<T> first, IEnumerable<T> second, IEqualityComparer<T> comparer = null)
         {
             if (first == null) throw new ArgumentNullException("first");
             if (second == null) throw new ArgumentNullException("second");
 
+            comparer = comparer ?? EqualityComparer<T>.Default;
+
             using (var firstIter = first.Reverse().GetEnumerator())
```

Comment:
How about the first approach suggested in this comment?
https://github.com/dotnet/corefx/issues/4750#issuecomment-172285275

Optimize TakeLast() for IList sequences, with the following:
```
int i = source.Count - 1;
while (i >= 0)
{
  yield return source[i];
  int count = source.Count;
  if (--i >= count)
    i = count - 1;
}
```

Alternatively, throw an exception if the source changes, like foreach does?

Then EndsWith() could follow @fsateler's suggestion of using TakeLast(), optimized for IList.